File: src/eval/metrics.ts
import type { Database } from "bun:sqlite";
import type { RetrievalEventRow, CompactionLogRow } from "../types.js";
import { writeCandidates, type MissContext } from "./benchmark-candidates.js";

export interface TierMetrics {
  precision: number;
  hitRate: number;
  avgScore: number;
  count: number;
}

export interface ContentTypeMetrics {
  precision: number;
  hitRate: number;
  count: number;
}

export interface MissEntry {
  query: string;
  topScore: number | null;
  timestamp: number;
}

export interface CompactionHealthMetrics {
  totalCompactions: number;
  avgPreservationRatio: number | null;
  entriesWithDrift: number;
}

export interface Metrics {
  precision: number;
  hitRate: number;
  missRate: number;
  mrr: number;
  avgLatencyMs: number;
  totalEvents: number;
  byTier: Record<string, TierMetrics>;
  byContentType: Record<string, ContentTypeMetrics>;
  topMisses: MissEntry[];
  falsePositives: MissEntry[];
  compactionHealth: CompactionHealthMetrics;
}

function computeGroupMetrics(
  events: RetrievalEventRow[],
): { precision: number; hitRate: number; avgScore: number } {
  const withOutcome = events.filter((e) => e.outcome_used !== null);
  const used = withOutcome.filter((e) => e.outcome_used === 1);
  const precision = withOutcome.length > 0 ? used.length / withOutcome.length : 0;

  const hitEvents = events.filter((e) => e.outcome_used !== null && e.outcome_used === 1);
  const eventsWithOutcome = events.filter((e) => e.outcome_used !== null);
  const hitRate =
    eventsWithOutcome.length > 0 ? hitEvents.length / eventsWithOutcome.length : 0;

  const scoresWithValue = events.filter((e) => e.top_score !== null);
  const avgScore =
    scoresWithValue.length > 0
      ? scoresWithValue.reduce((sum, e) => sum + (e.top_score as number), 0) /
        scoresWithValue.length
      : 0;

  return { precision, hitRate, avgScore };
}

export function computeMetrics(
  events: RetrievalEventRow[],
  similarityThreshold: number,
  compactionRows: CompactionLogRow[] = [],
  db?: Database,
): Metrics {
  if (events.length === 0) {
    return {
      precision: 0,
      hitRate: 0,
      missRate: 0,
      mrr: 0,
      avgLatencyMs: 0,
      totalEvents: 0,
      byTier: {},
      byContentType: {},
      topMisses: [],
      falsePositives: [],
      compactionHealth: computeCompactionHealth(compactionRows),
    };
  }

  // Precision: used / total among events with outcome
  const withOutcome = events.filter((e) => e.outcome_used !== null);
  const usedCount = withOutcome.filter((e) => e.outcome_used === 1).length;
  const precision = withOutcome.length > 0 ? usedCount / withOutcome.length : 0;

  // Hit rate: events where outcome_used = 1 / events with outcome
  const hitRate = withOutcome.length > 0 ? usedCount / withOutcome.length : 0;

  // Miss rate: events with top_score < threshold / total events
  const missCount = events.filter(
    (e) => e.top_score === null || e.top_score < similarityThreshold,
  ).length;
  const missRate = missCount / events.length;

  // MRR: simplified — 1.0 for used top results, 0 otherwise
  const mrrSum = withOutcome.reduce((sum, e) => {
    return sum + (e.outcome_used === 1 ? 1.0 : 0);
  }, 0);
  const mrr = withOutcome.length > 0 ? mrrSum / withOutcome.length : 0;

  // Average latency
  const latencies = events.filter((e) => e.latency_ms !== null);
  const avgLatencyMs =
    latencies.length > 0
      ? latencies.reduce((sum, e) => sum + (e.latency_ms as number), 0) / latencies.length
      : 0;

  // Group by tier
  const tierMap = new Map<string, RetrievalEventRow[]>();
  for (const e of events) {
    if (e.top_tier) {
      const group = tierMap.get(e.top_tier) ?? [];
      group.push(e);
      tierMap.set(e.top_tier, group);
    }
  }

  const byTier: Record<string, TierMetrics> = {};
  for (const [tier, group] of tierMap) {
    const { precision: p, hitRate: h, avgScore } = computeGroupMetrics(group);
    byTier[tier] = { precision: p, hitRate: h, avgScore, count: group.length };
  }

  // Group by content type
  const ctMap = new Map<string, RetrievalEventRow[]>();
  for (const e of events) {
    if (e.top_content_type) {
      const group = ctMap.get(e.top_content_type) ?? [];
      group.push(e);
      ctMap.set(e.top_content_type, group);
    }
  }

  const byContentType: Record<string, ContentTypeMetrics> = {};
  for (const [ct, group] of ctMap) {
    const { precision: p, hitRate: h } = computeGroupMetrics(group);
    byContentType[ct] = { precision: p, hitRate: h, count: group.length };
  }

  // Top misses: queries with lowest scores
  const topMisses: MissEntry[] = events
    .filter((e) => e.top_score === null || e.top_score < similarityThreshold)
    .sort((a, b) => (a.top_score ?? -1) - (b.top_score ?? -1))
    .slice(0, 10)
    .map((e) => ({ query: e.query_text, topScore: e.top_score, timestamp: e.timestamp }));

  // Write miss candidates for evolving benchmarks
  if (db && topMisses.length > 0) {
    try {
      const missContexts: MissContext[] = topMisses.map((miss) => {
        const event = events.find((e) => e.query_text === miss.query);
        let topEntryId: string | null = null;
        if (event) {
          const results = JSON.parse(event.results) as Array<{ entry_id: string }>;
          topEntryId = results[0]?.entry_id ?? null;
        }
        return { query: miss.query, topContent: null, topEntryId };
      });
      writeCandidates(db, topMisses, missContexts);
    } catch {
      // Don't fail metrics if candidate write fails
    }
  }

  // False positives: high score but outcome_used = 0
  const falsePositives: MissEntry[] = events
    .filter((e) => e.outcome_used === 0 && e.top_score !== null && e.top_score >= similarityThreshold)
    .sort((a, b) => (b.top_score ?? 0) - (a.top_score ?? 0))
    .slice(0, 10)
    .map((e) => ({ query: e.query_text, topScore: e.top_score, timestamp: e.timestamp }));

  return {
    precision,
    hitRate,
    missRate,
    mrr,
    avgLatencyMs,
    totalEvents: events.length,
    byTier,
    byContentType,
    topMisses,
    falsePositives,
    compactionHealth: computeCompactionHealth(compactionRows),
  };
}

export interface QueryDiffEntry {
  queryText: string;
  beforeOutcome: "used" | "unused" | "missing";
  afterOutcome: "used" | "unused" | "missing";
  beforeScore: number | null;
  afterScore: number | null;
}

export interface ComparisonResult {
  before: Metrics;
  after: Metrics;
  deltas: {
    precision: number;
    hitRate: number;
    mrr: number;
    missRate: number;
    avgLatencyMs: number;
  };
  byTier: Record<string, {
    before: TierMetrics;
    after: TierMetrics;
    deltas: { precision: number; hitRate: number; avgScore: number };
  }>;
  byContentType: Record<string, {
    before: ContentTypeMetrics;
    after: ContentTypeMetrics;
    deltas: { precision: number; hitRate: number };
  }>;
  queryDiff: {
    regressions: QueryDiffEntry[];
    improvements: QueryDiffEntry[];
  };
  warning?: string;
}

export function computeComparisonMetrics(
  eventsBefore: RetrievalEventRow[],
  eventsAfter: RetrievalEventRow[],
  similarityThreshold: number,
): ComparisonResult {
  const before = computeMetrics(eventsBefore, similarityThreshold);
  const after = computeMetrics(eventsAfter, similarityThreshold);

  const deltas = {
    precision: after.precision - before.precision,
    hitRate: after.hitRate - before.hitRate,
    mrr: after.mrr - before.mrr,
    missRate: after.missRate - before.missRate,
    avgLatencyMs: after.avgLatencyMs - before.avgLatencyMs,
  };

  // Per-tier deltas
  const allTiers = new Set([...Object.keys(before.byTier), ...Object.keys(after.byTier)]);
  const byTier: ComparisonResult["byTier"] = {};
  const emptyTier: TierMetrics = { precision: 0, hitRate: 0, avgScore: 0, count: 0 };
  for (const tier of allTiers) {
    const b = before.byTier[tier] ?? emptyTier;
    const a = after.byTier[tier] ?? emptyTier;
    byTier[tier] = {
      before: b,
      after: a,
      deltas: {
        precision: a.precision - b.precision,
        hitRate: a.hitRate - b.hitRate,
        avgScore: a.avgScore - b.avgScore,
      },
    };
  }

  // Per-content-type deltas
  const allTypes = new Set([...Object.keys(before.byContentType), ...Object.keys(after.byContentType)]);
  const byContentType: ComparisonResult["byContentType"] = {};
  const emptyType: ContentTypeMetrics = { precision: 0, hitRate: 0, count: 0 };
  for (const ct of allTypes) {
    const b = before.byContentType[ct] ?? emptyType;
    const a = after.byContentType[ct] ?? emptyType;
    byContentType[ct] = {
      before: b,
      after: a,
      deltas: {
        precision: a.precision - b.precision,
        hitRate: a.hitRate - b.hitRate,
      },
    };
  }

  // Query-level diff
  const beforeByQuery = new Map<string, RetrievalEventRow>();
  for (const e of eventsBefore) beforeByQuery.set(e.query_text, e);
  const afterByQuery = new Map<string, RetrievalEventRow>();
  for (const e of eventsAfter) afterByQuery.set(e.query_text, e);

  const regressions: QueryDiffEntry[] = [];
  const improvements: QueryDiffEntry[] = [];

  const allQueries = new Set([...beforeByQuery.keys(), ...afterByQuery.keys()]);
  for (const q of allQueries) {
    const b = beforeByQuery.get(q);
    const a = afterByQuery.get(q);
    const bOutcome = !b ? "missing" : b.outcome_used === 1 ? "used" : "unused";
    const aOutcome = !a ? "missing" : a.outcome_used === 1 ? "used" : "unused";

    if (bOutcome === aOutcome) continue;

    const entry: QueryDiffEntry = {
      queryText: q,
      beforeOutcome: bOutcome,
      afterOutcome: aOutcome,
      beforeScore: b?.top_score ?? null,
      afterScore: a?.top_score ?? null,
    };

    if (bOutcome === "used" && aOutcome !== "used") regressions.push(entry);
    if (aOutcome === "used" && bOutcome !== "used") improvements.push(entry);
  }

  let warning: string | undefined;
  if (eventsBefore.length === 0 || eventsAfter.length === 0) {
    warning = "Comparison requires retrieval events from both snapshots. One side has no data — metrics may not be meaningful.";
  }

  return { before, after, deltas, byTier, byContentType, queryDiff: { regressions, improvements }, warning };
}

function computeCompactionHealth(rows: CompactionLogRow[]): CompactionHealthMetrics {
  const withRatio = rows.filter((r) => r.preservation_ratio !== null);
  const withDrift = rows.filter((r) => r.semantic_drift !== null && r.semantic_drift > 0.2);

  return {
    totalCompactions: rows.length,
    avgPreservationRatio: withRatio.length > 0
      ? withRatio.reduce((sum, r) => sum + (r.preservation_ratio as number), 0) / withRatio.length
      : null,
    entriesWithDrift: withDrift.length,
  };
}

File: src/TotalRecall.Infrastructure/Eval/Metrics.cs
// src/TotalRecall.Infrastructure/Eval/Metrics.cs
//
// Plan 5 Task 5.3a — port of src-ts/eval/metrics.ts (the read/compute side
// only). Pure math over the existing RetrievalEventRow + CompactionAnalyticsRow
// projections from TotalRecall.Infrastructure.Telemetry. No I/O, no
// candidate-write side effects (Task 5.3b owns the write path on top of
// these read shapes).
//
// Faithful to the TS shapes:
//   - precision / hitRate / missRate / mrr / avgLatencyMs / totalEvents
//   - per-tier and per-content-type group breakdowns
//   - top misses + false positives (each as a ranked MissEntry list)
//   - compaction health derived from preservation_ratio + semantic_drift
//
// The TS file ALSO contains computeComparisonMetrics — that comparison logic
// is intentionally NOT ported here. Task 5.3b owns the compare command and
// will land it then. Documented at the bottom of this file.
//
// CompactionAnalyticsRow is defined in this file (TS keeps it in types.ts) because
// nothing under Infrastructure/Telemetry/CompactionLog.cs currently exposes a
// "list rows for analytics" projection — we add a minimal reader extension on
// CompactionLog (see CompactionLog.GetAllForAnalytics) and shape rows into
// this DTO before handing them to Compute.

using System;
using System.Collections.Generic;
using System.Linq;
using TotalRecall.Infrastructure.Telemetry;

namespace TotalRecall.Infrastructure.Eval;

/// <summary>Per-tier breakdown of precision/hitRate/avgScore.</summary>
public sealed record TierMetrics(
    double Precision,
    double HitRate,
    double AvgScore,
    int Count);

/// <summary>Per-content-type breakdown.</summary>
public sealed record ContentTypeMetrics(
    double Precision,
    double HitRate,
    int Count);

/// <summary>Single miss / false-positive entry.</summary>
public sealed record MissEntry(
    string Query,
    double? TopScore,
    long Timestamp);

/// <summary>Compaction health summary derived from compaction_log rows.</summary>
public sealed record CompactionHealthMetrics(
    int TotalCompactions,
    double? AvgPreservationRatio,
    int EntriesWithDrift);

/// <summary>
/// Aggregate retrieval-quality metrics computed by <see cref="Metrics.Compute"/>.
/// All numeric fields are zeroed when there are no events to score.
/// </summary>
public sealed record MetricsReport(
    double Precision,
    double HitRate,
    double MissRate,
    double Mrr,
    double AvgLatencyMs,
    int TotalEvents,
    IReadOnlyDictionary<string, TierMetrics> ByTier,
    IReadOnlyDictionary<string, ContentTypeMetrics> ByContentType,
    IReadOnlyList<MissEntry> TopMisses,
    IReadOnlyList<MissEntry> FalsePositives,
    CompactionHealthMetrics CompactionHealth);

/// <summary>
/// Pure-math metric aggregator. Read-only — produces a <see cref="MetricsReport"/>
/// from a list of retrieval events plus optional compaction rows. No DB,
/// no filesystem, no logging. Mirrors <c>computeMetrics</c> in
/// <c>src-ts/eval/metrics.ts</c>.
/// </summary>
public static class Metrics
{
    /// <summary>
    /// Compute the aggregate metrics report. Empty events produce a zeroed
    /// report (still includes compaction health if rows are supplied).
    /// </summary>
    public static MetricsReport Compute(
        IReadOnlyList<RetrievalEventRow> events,
        double similarityThreshold,
        long nowMs,
        long graceWindowMs,
        IReadOnlyList<CompactionAnalyticsRow>? compactionRows = null)
    {
        ArgumentNullException.ThrowIfNull(events);
        var compaction = compactionRows ?? Array.Empty<CompactionAnalyticsRow>();

        if (events.Count == 0)
        {
            return new MetricsReport(
                Precision: 0,
                HitRate: 0,
                MissRate: 0,
                Mrr: 0,
                AvgLatencyMs: 0,
                TotalEvents: 0,
                ByTier: new Dictionary<string, TierMetrics>(),
                ByContentType: new Dictionary<string, ContentTypeMetrics>(),
                TopMisses: Array.Empty<MissEntry>(),
                FalsePositives: Array.Empty<MissEntry>(),
                CompactionHealth: ComputeCompactionHealth(compaction));
        }

        // Resolve each event's outcome at read time; pending (recent NULL) events
        // are excluded from every rate.
        var scored = events
            .Select(e => (Event: e, Outcome: ResolveOutcome(e, nowMs, graceWindowMs)))
            .Where(x => x.Outcome.HasValue)
            .Select(x => (x.Event, Used: x.Outcome!.Value))
            .ToList();

        var hits = scored.Count(x => x.Used);
        var precision = scored.Count > 0 ? (double)hits / scored.Count : 0.0;
        var hitRate = precision;          // mirrors the prior identity
        var mrr = precision;              // simplified MRR on the same hit/miss basis

        // Miss rate (score quality) over the same scored set.
        var missCount = scored.Count(x =>
            x.Event.TopScore is null || x.Event.TopScore.Value < similarityThreshold);
        var missRate = scored.Count > 0 ? (double)missCount / scored.Count : 0.0;

        // Average latency over scored events that recorded one.
        var latencies = scored.Where(x => x.Event.LatencyMs.HasValue).ToList();
        var avgLatencyMs = latencies.Count > 0
            ? latencies.Average(x => (double)x.Event.LatencyMs!.Value)
            : 0.0;

        // Group by tier / content type over scored events, carrying the resolved outcome.
        var byTier = scored
            .Where(x => !string.IsNullOrEmpty(x.Event.TopTier))
            .GroupBy(x => x.Event.TopTier!, StringComparer.Ordinal)
            .ToDictionary(
                g => g.Key,
                g =>
                {
                    var (p, h, avg) = ComputeGroupMetrics(g);
                    return new TierMetrics(p, h, avg, g.Count());
                },
                StringComparer.Ordinal);

        var byContentType = scored
            .Where(x => !string.IsNullOrEmpty(x.Event.TopContentType))
            .GroupBy(x => x.Event.TopContentType!, StringComparer.Ordinal)
            .ToDictionary(
                g => g.Key,
                g =>
                {
                    var (p, h, _) = ComputeGroupMetrics(g);
                    return new ContentTypeMetrics(p, h, g.Count());
                },
                StringComparer.Ordinal);

        // Top misses: lowest scoring scored events (null sorts as -1).
        var topMisses = scored
            .Where(x => x.Event.TopScore is null || x.Event.TopScore.Value < similarityThreshold)
            .OrderBy(x => x.Event.TopScore ?? -1.0)
            .Take(10)
            .Select(x => new MissEntry(x.Event.QueryText, x.Event.TopScore, x.Event.Timestamp))
            .ToList();

        // False positives: explicit negatives with a high score.
        var falsePositives = scored
            .Where(x => !x.Used && x.Event.TopScore.HasValue && x.Event.TopScore.Value >= similarityThreshold)
            .OrderByDescending(x => x.Event.TopScore ?? 0.0)
            .Take(10)
            .Select(x => new MissEntry(x.Event.QueryText, x.Event.TopScore, x.Event.Timestamp))
            .ToList();

        return new MetricsReport(
            Precision: precision,
            HitRate: hitRate,
            MissRate: missRate,
            Mrr: mrr,
            AvgLatencyMs: avgLatencyMs,
            TotalEvents: events.Count,
            ByTier: byTier,
            ByContentType: byContentType,
            TopMisses: topMisses,
            FalsePositives: falsePositives,
            CompactionHealth: ComputeCompactionHealth(compaction));
    }

    // Read-time outcome resolution. true = hit, false = miss, null = pending (excluded).
    private static bool? ResolveOutcome(RetrievalEventRow e, long nowMs, long graceWindowMs)
    {
        if (e.OutcomeUsed == true) return true;
        if (e.OutcomeUsed == false) return false;
        return e.Timestamp < nowMs - graceWindowMs ? false : (bool?)null;
    }

    private static (double Precision, double HitRate, double AvgScore) ComputeGroupMetrics(
        IEnumerable<(RetrievalEventRow Event, bool Used)> scored)
    {
        var list = scored as IList<(RetrievalEventRow Event, bool Used)> ?? scored.ToList();
        var used = list.Count(x => x.Used);
        var precision = list.Count > 0 ? (double)used / list.Count : 0.0;
        var withScore = list.Where(x => x.Event.TopScore.HasValue).ToList();
        var avgScore = withScore.Count > 0 ? withScore.Average(x => x.Event.TopScore!.Value) : 0.0;
        return (precision, precision, avgScore);
    }

    private static CompactionHealthMetrics ComputeCompactionHealth(
        IReadOnlyList<CompactionAnalyticsRow> rows)
    {
        var withRatio = rows.Where(r => r.PreservationRatio.HasValue).ToList();
        var withDrift = rows.Count(r =>
            r.SemanticDrift.HasValue && r.SemanticDrift.Value > 0.2);

        return new CompactionHealthMetrics(
            TotalCompactions: rows.Count,
            AvgPreservationRatio: withRatio.Count > 0
                ? withRatio.Average(r => r.PreservationRatio!.Value)
                : null,
            EntriesWithDrift: withDrift);
    }
}
